// styled-components/lib/test/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.expectCSSMatches = exports.resetNoParserStyled = exports.resetStyled = undefined;

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _styled2 = require('../constructors/styled');

var _styled3 = _interopRequireDefault(_styled2);

var _css = require('../constructors/css');

var _css2 = _interopRequireDefault(_css);

var _constructWithOptions2 = require('../constructors/constructWithOptions');

var _constructWithOptions3 = _interopRequireDefault(_constructWithOptions2);

var _StyleSheet = require('../models/StyleSheet');

var _StyleSheet2 = _interopRequireDefault(_StyleSheet);

var _flatten = require('../utils/flatten');

var _flatten2 = _interopRequireDefault(_flatten);

var _stringifyRules = require('../utils/stringifyRules');

var _stringifyRules2 = _interopRequireDefault(_stringifyRules);

var _StyledComponent2 = require('../models/StyledComponent');

var _StyledComponent3 = _interopRequireDefault(_StyledComponent2);

var _ComponentStyle2 = require('../models/ComponentStyle');

var _ComponentStyle3 = _interopRequireDefault(_ComponentStyle2);

var _css3 = require('../no-parser/css');

var _css4 = _interopRequireDefault(_css3);

var _flatten3 = require('../no-parser/flatten');

var _flatten4 = _interopRequireDefault(_flatten3);

var _stringifyRules3 = require('../no-parser/stringifyRules');

var _stringifyRules4 = _interopRequireDefault(_stringifyRules3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* Ignore hashing, just return class names sequentially as .a .b .c etc */

/**
 * This sets up our end-to-end test suite, which essentially makes sure
 * our public API works the way we promise/want
 */
var index = 0;
var classNames = function classNames() {
  return String.fromCodePoint(97 + index++);
};

var resetStyled = exports.resetStyled = function resetStyled() {
  _StyleSheet2.default.reset();
  index = 0;

  var ComponentStyle = (0, _ComponentStyle3.default)(classNames, _flatten2.default, _stringifyRules2.default);
  var constructWithOptions = (0, _constructWithOptions3.default)(_css2.default);
  var StyledComponent = (0, _StyledComponent3.default)(ComponentStyle, constructWithOptions);

  return (0, _styled3.default)(StyledComponent, constructWithOptions);
};

var resetNoParserStyled = exports.resetNoParserStyled = function resetNoParserStyled() {
  _StyleSheet2.default.reset();
  index = 0;

  var ComponentStyle = (0, _ComponentStyle3.default)(classNames, _flatten4.default, _stringifyRules4.default);
  var constructWithOptions = (0, _constructWithOptions3.default)(_css4.default);
  var StyledComponent = (0, _StyledComponent3.default)(ComponentStyle, constructWithOptions);

  return (0, _styled3.default)(StyledComponent, constructWithOptions);
};

var stripWhitespace = function stripWhitespace(str) {
  return str.trim().replace(/([;\{\}])/g, '$1  ').replace(/\s+/g, ' ');
};
var expectCSSMatches = exports.expectCSSMatches = function expectCSSMatches(expectation) {
  var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : { ignoreWhitespace: true };

  var css = _StyleSheet2.default.getCSS({ min: false });
  if (opts.ignoreWhitespace) {
    (0, _expect2.default)(stripWhitespace(css)).toEqual(stripWhitespace(expectation));
  } else {
    (0, _expect2.default)(css).toEqual(expectation);
  }
  return css;
};
